airflow/jobs.py

```diff
@@ -1891,112 +1940,126 @@ def _manage_executor_state(self, started):
-                # all tasks part of the backfill are scheduled to run
-                if ti.state == State.NONE:
-                    ti.set_state(State.SCHEDULED, session=session)
-                tasks_to_run[ti.key] = ti
+        for ti in dag_run.get_task_instances():
+            # all tasks part of the backfill are scheduled to run
+            if ti.state == State.NONE:
+                ti.set_state(State.SCHEDULED, session=session)
+            tasks_to_run[ti.key] = ti
+
+        return tasks_to_run
+
+    def _log_progress(self, ti_status, tasks_to_run):
+        msg = ' | '.join([
+            "[backfill progress]",
+            "finished run {0} of {1}",
+            "tasks waiting: {2}",
+            "succeeded: {3}",
+            "kicked_off: {4}",
+            "failed: {5}",
+            "skipped: {6}",
+            "deadlocked: {7}",
+            "not ready: {8}"
+        ]).format(
+            ti_status.finished_runs,
+            ti_status.total_runs,
+            len(tasks_to_run),
+            len(ti_status.succeeded),
+            len(ti_status.started),
+            len(ti_status.failed),
+            len(ti_status.skipped),
+            len(ti_status.deadlocked),
+            len(ti_status.not_ready))
+        self.logger.info(msg)
+
+        self.logger.debug("Finished dag run loop iteration. "
+                          "Remaining tasks {}"
+                          .format(tasks_to_run.values()))
 
-            next_run_date = self.dag.following_schedule(next_run_date)
+    @provide_session
+    def _process_backfill_task_instances(self, tasks_to_run,
+                                         active_dag_runs,
+                                         ti_status,
+                                         executor,
+                                         pickle_id,
+                                         start_date=None, session=None):
+        """
+        Process a set of task instances from a set of dag runs. Special handling is done
+        to account for different task instance states that could be present when running
+        them in a backfill process.
+        :param tasks_to_run: the tasks to run in this Backfill job
+        :type tasks_to_run: dict
+        :param status: the internal status of the job
+        :type status: BackfillJob._DagRunTaskStatus
+        :param executor: the executor to run the task instances
+        :param pickle_id: the pickle_id if dag is pickled, None otherwise
+        :param start_date: the start date of the backfill job
+        :param session: the current session object
+        """
 
-        finished_runs = 0
-        total_runs = len(active_dag_runs)
+        executed_run_dates = []
 
-        # Triggering what is ready to get triggered
-        while (len(tasks_to_run) > 0 or len(started) > 0) and not deadlocked:
+        while ((len(tasks_to_run) > 0 or len(ti_status.started) > 0) and
+                not ti_status.deadlocked):
```

Comment:
Let's make the ti_status.deadlocked consistent with the other checks in this while loop while we are refactoring it (> 0)